src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs

```diff
@@ -126,69 +101,37 @@ public static StringComparerBase Pick(ReadOnlySpan<string> uniqueStrings, bool i
                             {
-                                if (!IsAllAscii(ss.AsSpan(ss.Length + rightComparer.Index, rightComparer.Count)))
-                                {
-                                    // keep the slower non-ascii comparer since we have some non-ascii text
-                                    return rightComparer;
-                                }
+                                results = new(false, ignoreCase, 0, 0, rightComparer.Index, rightComparer.Count);
+                                return true;
                             }
-
-                            // optimize for all-ascii case
-                            return new RightJustifiedCaseInsensitiveAsciiSubstringComparer
-                            {
-                                Index = rightComparer.Index,
-                                Count = rightComparer.Count,
-                            };
-                        }
-
-                        // Optimize the single char case
-                        if (rightComparer.Count == 1)
-                        {
-                            return new RightJustifiedSingleCharComparer
-                            {
-                                Index = rightComparer.Index,
-                                Count = 1,
-                            };
                         }
 
-                        return rightComparer;
+                        results = new(true, ignoreCase, 0, 0, rightComparer.Index, rightComparer.Count);
+                        return true;
                     }
                 }
             }
 
-            return null;
+            results = default;
+            return false;
         }
 
-        private static StringComparerBase PickFullStringComparer(ReadOnlySpan<string> uniqueStrings, bool ignoreCase)
+        private static void UseFullString(ReadOnlySpan<string> uniqueStrings, bool ignoreCase, out AnalysisResults results)
         {
-            if (!ignoreCase)
-            {
-                return new FullStringComparer();
-            }
-
             foreach (string s in uniqueStrings)
             {
                 if (!IsAllAscii(s.AsSpan()))
                 {
-                    return new FullCaseInsensitiveStringComparer();
+                    results = new(false, ignoreCase, 0, 0, 0, 0);
+                    return;
                 }
             }
 
-            return new FullCaseInsensitiveAsciiStringComparer();
-        }
-
-        private sealed class ComparerWrapper : IEqualityComparer<string>
-        {
-            private readonly SubstringComparerBase _comp;
-
-            public ComparerWrapper(SubstringComparerBase comp) => _comp = comp;
-
-            public bool Equals(string? x, string? y) => _comp.EqualsPartial(x, y);
-            public int GetHashCode([DisallowNull] string obj) => _comp.GetHashCode(obj);
+            results = new(true, ignoreCase, 0, 0, 0, 0);
```

Comment:
Same feedback about all of these we could consolidate into a single results setting using a bool for the first arg.